xtask/src/test_runner.rs

```diff
@@ -0,0 +1,904 @@
+                out_clone2.lock().unwrap().push_str(&line);
+                out_clone2.lock().unwrap().push('\n');
+            }
+        });
+
+        Ok((child, output, stdout_handle, stderr_handle))
+    }
+
+    /// Runs a command with a timeout, returning the exit code
+    pub fn run_command_with_timeout(&self, cmd: &mut Command, timeout: Duration) -> Result<i32> {
+        log::debug!("Running command: {cmd:?}");
+        self.setup_command(cmd);
+
+        let mut child = cmd.spawn()?;
+        let completed = Arc::new(AtomicBool::new(false));
+        let child_id = child.id();
+        let completed_clone = Arc::clone(&completed);
+
+        let timer = thread::spawn(move || {
+            let interval = Duration::from_millis(100);
+            let mut elapsed = Duration::ZERO;
+
+            while elapsed < timeout {
+                thread::sleep(interval);
+                elapsed += interval;
+                if completed_clone.load(Ordering::SeqCst) {
+                    return;
+                }
+            }
+
+            log::warn!("Command timed out after {timeout:?}, killing process {child_id}");
+            Self::terminate_process(&mut None);
+        });
+
+        let status = match child.wait() {
+            Ok(s) => {
+                completed.store(true, Ordering::SeqCst);
+                s
+            }
+            Err(e) => {
+                completed.store(true, Ordering::SeqCst);
+                thread::sleep(Duration::from_millis(10));
+                return Err(format!("Command execution failed: {e}").into());
+            }
+        };
+
+        let _ = timer.join();
+        let exit_code = status.code().unwrap_or(1);
+        log::debug!("Command exit code: {exit_code}");
+        Ok(exit_code)
+    }
+
+    /// Runs a command for a specified duration, returning whether it terminated
+    /// naturally
+    pub fn run_command_for(&self, cmd: &mut Command, duration: Duration) -> Result<bool> {
+        log::debug!("Running command: {cmd:?}");
+        let mut child = cmd.spawn()?;
+        let start_time = Instant::now();
+        let mut naturally_terminated = false;
+
+        if let Ok(Some(_)) = child.try_wait() {
+            naturally_terminated = true;
+        } else {
+            thread::sleep(duration);
+            if let Ok(Some(_)) = child.try_wait() {
+                naturally_terminated = true;
+            }
+        }
+
+        if !naturally_terminated {
+            log::info!(
+                "Command ran for {:?}, terminating process {}",
+                start_time.elapsed(),
+                child.id()
+            );
+            Self::terminate_process(&mut Some(&mut child));
+        }
+
+        Self::restore_terminal();
+        log::debug!("Command completed after {:?}", start_time.elapsed());
+
+        Ok(naturally_terminated)
+    }
+
+    fn create_espflash_command(&self, args: &[&str]) -> Command {
+        let mut cmd = Command::new("espflash");
```

Comment:
This tests the system installed espflash, not the changes in a given PR/local working tree